### ISSUE TITLE
Resolving Android Dependencies: Job failed: temp file/dir is not empty

### ISSUE THREAD

**Issue by sampenguin** [googlesamples/unity-jar-resolver#198]
I'm getting this exception constantly when PlayServicesResolver attempts to run. I'll see a few popup windows that say Resolving Android Dependencies, Unity will chug for a few seconds, and then I get left with a tiny progress bar window that never goes away and this exception in the log. It happens right after "Assets\PlayServicesResolver\Editor\play-services-resolver.txt metadata will be checked" shows in the console window.

Any time I make a change requiring a recompile, it goes through this process again, which basically grinds my development to a halt. Have tried restarting Unity, deleting Temp dir contents, but it's apparently something that Resolving Android Dependencies generates on its own. Any ideas on a workaround or what may be causing this?

Happens in Unity 2018.3.9f1 and Unity 2018.3.1f1. I have GoogleMobileAds and GooglePlayGames in this project, managed via EasyMobilePro.

```
Job failed with exception: System.IO.IOException: Directory C:\<unity_project>\Temp\e8udtmqg.5bz is not empty
  at System.IO.Directory.Delete (System.String path) [0x0006f] in <d7ac571ca2d04b2f981d0d886fa067cf>:0 
  at Google.FileUtils.DeleteExistingFileOrDirectory (System.String path, System.Boolean includeMetaFiles) [0x00077] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/FileUtils.cs:71 
  at GooglePlayServices.ResolverVer1_1.ShouldExplode (System.String aarPath) [0x00242] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/ResolverVer1_1.cs:1518 
  at GooglePlayServices.ResolverVer1_1+<GradleResolution>c__AnonStorey18.<>m__2F (GooglePlayServices.CommandLine+Result result) [0x000b5] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/ResolverVer1_1.cs:648 
  at GooglePlayServices.ResolverVer1_1+<GradleResolution>c__AnonStorey18+<GradleResolution>c__AnonStorey1A.<>m__3A () [0x00000] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/ResolverVer1_1.cs:710 
  at Google.RunOnMainThread.ExecuteNext () [0x0003d] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:363 
UnityEngine.Debug:LogError(Object)
Google.RunOnMainThread:ExecuteNext() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:365)
Google.RunOnMainThread:<ExecuteAll>m__16() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:394)
Google.RunOnMainThread:RunAction(Action) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:234)
Google.RunOnMainThread:ExecuteAll() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:392)
Google.RunOnMainThread:Run(Action, Boolean) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:349)
GooglePlayServices.<GradleResolution>c__AnonStorey18:<>m__30(Result) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/ResolverVer1_1.cs:711)
GooglePlayServices.<GradleResolution>c__AnonStorey18:<>m__31(Result) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/ResolverVer1_1.cs:762)
GooglePlayServices.ProgressReporter:SignalComplete() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/CommandLineDialog.cs:116)
GooglePlayServices.ProgressReporter:CommandLineToolCompletion(Result) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/CommandLineDialog.cs:107)
GooglePlayServices.<RunAsync>c__AnonStorey2:<>m__4() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/CommandLine.cs:136)
Google.RunOnMainThread:ExecuteNext() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:363)'
Google.RunOnMainThread:<ExecuteAll>m__16() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:394)
Google.RunOnMainThread:RunAction(Action) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:234)
Google.RunOnMainThread:ExecuteAll() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:392)
UnityEditor.EditorApplication:Internal_CallUpdateFunctions()
```

**Comment by sampenguin** [comment]
After battling this for days and posting this issue, then I just found a workaround/solution.

I removed PlayServicesResolver from my Unity project, then reimported the package. This time, I got a different exception:

```

Stopped polling job due to exception: System.IO.IOException: Sharing violation on path Assets/Plugins/Android/com.android.support.support-vector-drawable-26.1.0.aar
  at System.IO.File.Delete (System.String path) [0x00073] in <d7ac571ca2d04b2f981d0d886fa067cf>:0 
  at Google.FileUtils.DeleteExistingFileOrDirectory (System.String path, System.Boolean includeMetaFiles) [0x0009e] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/FileUtils.cs:76 
  at GooglePlayServices.PlayServicesResolver.DeleteFiles (System.Collections.Generic.IEnumerable`1[T] filenames) [0x0001d] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1190 
  at GooglePlayServices.PlayServicesResolver.DeleteLabeledAssets () [0x00000] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1603 
  at GooglePlayServices.PlayServicesResolver.ResolveUnsafe (System.Action`1[T] resolutionComplete, System.Boolean forceResolution, System.Boolean isAutoResolveJob) [0x001b5] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1418 
  at GooglePlayServices.PlayServicesResolver+<ScheduleResolve>c__AnonStorey12.<>m__22 () [0x00000] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1305 
  at GooglePlayServices.PlayServicesResolver.ExecuteNextResolveJob () [0x00069] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1224 
  at GooglePlayServices.PlayServicesResolver.ScheduleResolve (System.Boolean forceResolution, System.Action`1[T] resolutionCompleteWithResult, System.Boolean isAutoResolveJob) [0x000a0] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:1316 
  at GooglePlayServices.PlayServicesResolver.AutoResolve (System.Action resolutionComplete) [0x0001c] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:906 
  at GooglePlayServices.PlayServicesResolver.<ScheduleAutoResolve>m__27 () [0x00016] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/PlayServicesResolver/src/PlayServicesResolver.cs:885 
  at Google.RunOnMainThread.ExecutePollingJobs () [0x0004a] in /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:286 
UnityEngine.Debug:LogError(Object)
Google.RunOnMainThread:ExecutePollingJobs() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:289)
Google.RunOnMainThread:<ExecuteAll>m__16() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:400)
Google.RunOnMainThread:RunAction(Action) (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:234)
Google.RunOnMainThread:ExecuteAll() (at /usr/local/google/home/cleoschneider/google-play-resolver/unity-jar-resolver/source/VersionHandlerImpl/src/RunOnMainThread.cs:392)
UnityEditor.EditorApplication:Internal_CallUpdateFunctions()
```

So then I deleted that "com.android.support.support-vector-drawable-26.1.0.aar" file, and kicked off another compile to initiate the resolver again. This time resolving android dependencies completed, and now I see about 20 new files in my Plugins/Android directory. 

My guess is that aar was corrupted somehow, and the resolver got hung up on it in some way. It was redownloaded as part of the resolution.

**Comment by sampenguin** [comment]
Scratch that. It's intermittent, still a problem. It fails randomly with attempted deletion of randomly generated named dirs in the Temp dir, or sharing violations on random .aar files. Guess I just got lucky that one time.

If it's useful, I'm on an SSD and pretty fast multicore rig (Threadripper). I'm wondering if this is a multithreaded process if something is happening too fast in IO...? Stab in the dark but I can't think what would be causing this behavior.

**Comment by a-maurice** [comment]
Hi @sampenguin 

Definitely sounds like something odd is happening during it.  If you can, could in the toolbar go to Assets > Play Services Resolver > Android Resolver > Settings, enable "Verbose Logging", and share that?  It will become a lot chattier, but that might help identify why it is getting hung up where it is.  Also, if it is blocking your development, you can disable auto-resolution, but that will likely prevent your Android builds from working, since it won't be grabbing the necessary aars.

**Comment by sampenguin** [comment]
OK.
I see this towards the end:

- Command completed: Successfully executed '<project_path>\Temp\PlayServicesResolverGradle\gradlew.bat -b <lots of other stuff>

Then followed by a bunch of "Extracting <whatever>.aar to..." messages,

Then the "is not empty" exception as reported in original post. It's random which aar extraction it happens on, and sometimes it doesn't happen.

FYI, yesterday I also copied the whole project to a standard HDD with plenty of room (after updating SSD drivers and firmware to no effect), and it's still happening on the HDD *shrug*.

**Comment by a-maurice** [comment]
So, the last time this came up was https://github.com/googlesamples/unity-jar-resolver/issues/19.  The problem then seemed to be an anti-virus program preventing the deletion of those temporary files.  Do you think that might be causing your problem?

**Comment by sampenguin** [comment]
Funny you mention that, I just had that exact thought and went to look at my AV, but I've already had real time scanning disabled on my development directories on both the SSD and HDD for a long time. I can still try disabling it altogether and see if I can reproduce.

One other observation that might help to reproduce it: it *seems* like if I delete an aar file, then hit play in the editor (I do this to kick off a resolution because selecting Resolve and Force Resolve in the Assets->Play Services Resolver menu drop down both do nothing after getting the exception), then rapidly hit play again to stop in editor, it seems to happen more often. My last successful resolution I did with just hitting play once, letting the resolve finish (takes about 10-20 seconds), and then hitting play again to stop. So maybe something with Unity doing things in the Temp folder at the same time as the Resolver could be a contributing factor.

**Comment by sampenguin** [comment]
Debunked that theory. Randomly get the exception whether I let it try to finish before stopping play or not. 

Another odd thing, I've had two successful resolutions out of the last 10 tries. I think I have to delete everything in /Temp (except UnityLockFile) first, then let the resolver run without doing anything else to get success.

1) On the first successful resolution (log showed Resolution Succeeded and no exception), it left 3 exploded directories instead of .aar files in Plugins/Android/ (e.g. com.google.games.gpgs-plugin-support-0.9.62/ instead of the corresponding .aar file), while all the other aar files (~25 of them) still exist as singular files.

2) On the second successful resolution (log showed Resolution Succeeded and no exception), there were not any exploded directories, and the previously exploded 3 directories were now still there as .aar files.

???

**Comment by a-maurice** [comment]
Yeah, looking into it more, we can't seem to find anything that jumps out as the problem.  The logic that cleans up the Temp directory is here: https://github.com/googlesamples/unity-jar-resolver/blob/f8ba790ef8d19b3df8834ddb21dcb1ca457c2b63/source/VersionHandlerImpl/src/FileUtils.cs#L58

Anti-virus is still our first thought.  Another theory is that multiple resolutions are running in parallel.  The only way we can think something like that would happen would be if other libraries are calling the internal functions directly.  Do you see this problem when you only have one of GoogleMobileAds or GooglePlayGames?

**Comment by sampenguin** [comment]
AV protection is completely disabled, and I excluded every conceivable process and folder I could think of prior to just disabling it, and in all cases I still intermittently get the error. I also forced ownership of my current user with full control on all my development directories just to be sure it wasn't some weird permission thing (although if it was, I would expect the problem to be 100% of the time). Since it's random which file it happens on, or if it even happens at all, it seems like a race condition/timing issue to me.

Looking at the source you linked, and I came across this which sounds like it might fit as a problem on Windows anyway (see the 2nd most popular answer):
https://stackoverflow.com/questions/329355/cannot-delete-directory-with-directory-deletepath-true

> If you are trying to recursively delete directory a and directory a\b is open in Explorer, b will be deleted but you will get the error 'directory is not empty' for a even though it is empty when you go and look. The current directory of any application (including Explorer) retains a handle to the directory. When you call Directory.Delete(true), it deletes from bottom up: b, then a. If b is open in Explorer, Explorer will detect the deletion of b, change directory upwards cd .. and clean up open handles. Since the file system operates asynchronously, the Directory.Delete operation fails due to conflicts with Explorer.

So maybe if those delete calls were wrapped in try/catches and you just logged it in verbose mode, that would be enough to stop the job from dying in unity?
